bus-mapping/src/evm/opcodes/mload.rs

```diff
@@ -0,0 +1,88 @@
+use super::Opcode;
+use crate::{
+    evm::GlobalCounter,
+    exec_trace::ExecutionStep,
+    operation::{container::OperationContainer, MemoryOp, StackOp, RW},
+    Error,
+};
+use core::ops::Deref;
+
+/// Number of ops that MLOAD adds to the container & busmapping
+const MLOAD_OP_NUM: usize = 3;
+
+/// Placeholder structure used to implement [`Opcode`] trait over it corresponding to the
+/// [`OpcodeId::MLOAD`](crate::evm::OpcodeId::MLOAD) `OpcodeId`.
+/// This is responsible of generating all of the associated [`StackOp`]s and [`MemoryOp`]s and place them
+/// inside the trace's [`OperationContainer`].
+#[derive(Debug, Copy, Clone)]
+pub(crate) struct Mload;
+
+impl Opcode for Mload {
+    #[allow(unused_variables)]
+    fn gen_associated_ops(
+        &self,
+        exec_step: &mut ExecutionStep,
+        container: &mut OperationContainer,
+        next_steps: &[ExecutionStep],
+    ) -> Result<usize, Error> {
+        //
+        // First stack read
+        //
+        let stack_value_read = exec_step
+            .stack()
+            .deref()
+            .last()
+            .cloned()
+            .ok_or(Error::InvalidStackPointer)?;
+        let stack_position = exec_step.stack().last_filled();
+
+        // Manage first stack read at latest stack position
+        let stack_read = StackOp::new(
+            RW::READ,
+            GlobalCounter::from(exec_step.gc().0 + 1),
+            stack_position,
+            stack_value_read,
+        );
+
+        exec_step
+            .bus_mapping_instance_mut()
+            .push(container.insert(stack_read));
+
+        //
+        // First mem read
+        //
+        let last_mem_item = exec_step
+            .memory()
+            .deref()
+            .last()
+            .cloned()
+            .ok_or(Error::InvalidMemoryPointer)?;
+
+        // Read operation at memory address: stack_read.value
+        let mem_read = MemoryOp::new(
+            RW::READ,
+            GlobalCounter::from(exec_step.gc().0 + 2),
+            last_mem_item.0,
+            last_mem_item.1.clone(),
+        );
+
+        exec_step
+            .bus_mapping_instance_mut()
+            .push(container.insert(mem_read));
+
+        //
+        // First stack write
+        //
+        let stack_write = StackOp::new(
+            RW::WRITE,
+            GlobalCounter::from(exec_step.gc().0 + 3),
+            stack_position,
+            last_mem_item.1,
```

Comment:
Why is the word loaded from memory is the last one in current memroy? I suppose it should be taken from the last item of `next_steps[0]`.

Reply:
Because if you check the trace. The current step is the one which has the `MLOAD` opcode. That means that there we already have the value in the memory that we want to read. And the value won't change from the read in memory to the write in stack. 

Does it make sense?

Reply:
Yes we have the value in the memory that we want to read, but the position won't always be the last one, it should use the value popped from stack as memory address, right?

Reply:
Solved!

Reply:
I thought you would use the last item of `next_steps[0].stack` as `mem_read_value` directly, but getting from memory also works tho.

Reply:
That's indeed the case IIUC.
```
let stack_value_read = exec_step
            .stack()
            .deref()
            .last()
            .cloned()
            .ok_or(Error::InvalidStackPointer)?;
// .....
// .....
let mem_read_addr: MemoryAddress = stack_value_read.try_into()?;
let mem_read_value = next_steps[0].memory()[mem_read_addr].clone();
```

The address is the latest stack value read. And we read at this address for step `next_steps[0]`.

Reply:
Yes, but the `mem_read_value` should be pushed to stack right? So we should also be able to get it by:

```rust
let mem_read_value = next_steps[0].stack()
            .deref()
            .last()
            .cloned()
            .ok_or(Error::InvalidStackPointer)?
```